### PR TITLE
HPO-Wrapper for Gate-Level Calibration

### PR DESCRIPTION
**Pull request by lukasvoss**
This PR is for a Hyperparameter Optimization Wrapper allowing for automated HPO based on an input YAML file.

1.) The HPO class ``HyperparameterOptimizer`` takes as input

    - QEnvConfig
    - a path to the agent_config.yaml file
    - a path to where the best configuration of hyperparameters shall be saved as pickle file
    - a boolean whether to retrieve logging information of the HPO progress
    - and int specifying the number of HPO trials
The class automatically reads in the HPO information from the yaml file provided and then creates finds the best configuration of hyperparams. The number of trials can be both set in the yaml file and the class as input. If given as input, this number will overrule any input in the yaml.

2.) I added some input handling functions for the HPO in the helper_functions.py file.

3.) Before the PR, I created this new feature branch "test/rebase" and synced with the codebase with the main branch of Arthur's repo discarding all my prior commits since there were merge conflicts. I then added my changes for the HPO, etc.

### CODE REVIEW

**Comment by lukasvoss** [comment]
Resolved the merge conflicts for these files and respective conflicts:

1. ppo.py: Use a try-except block to catch errors in training. Error-causing trials will be caught in the HPO and it will be proceeded, the hyperparam configuration will be penalized and it will be proceeded with the next hyperparam configuration.
2. quantumenvironment.py: Change import statement for ComputeUncompute
3. q_env_config.py: default param vector for the CX gate calibration had a line break

File: quantumenvironment.py [hunk 1]
```diff
@@ -40,6 +40,7 @@
 )
 from qiskit.quantum_info.states import DensityMatrix, Statevector
 from qiskit.transpiler import Layout
```
Comment (arthurostrauss) [inline quantumenvironment.py:42]: You should not need this import anymore in principle

File: template_configurations/qiskit/q_env_config.py [hunk 2]
```diff
@@ -59,11 +60,10 @@ def apply_parametrized_circuit(
         optimal_params[5] + params[5],
         q_reg[1],
     )
-
     my_qc.rzx(optimal_params[6] + params[6], q_reg[0], q_reg[1])
-    # my_qc.u(np.pi *params[0], np.pi *params[1], np.pi *params[2], 0)
-    # my_qc.u(np.pi *params[3], np.pi *params[4], np.pi *params[5], 1)
-    # my_qc.rzx(np.pi * params[6], 0, 1)
+    # my_qc.u(2 * np.pi * params[0], 2 *  np.pi *params[1], 2 * np.pi * params[2], 0)
```
Comment (arthurostrauss) [inline template_configurations/qiskit/q_env_config.py:64]: If the action scales between -1 and 1, then pi should be enough.

**Review by arthurostrauss — CHANGES_REQUESTED** [review]
Very good job. Overall there are very minor modifications to be made. One big one I suggest is that you install black (pip install black), and then enter black . in your terminal when you're in the folder. this will trigger some reformatting of the code which is always good practise.

File: hyperparameter_optimization.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    def __init__(
+            self, 
+            gate_q_env_config: QEnvConfig, 
+            path_agent_config: str, 
+            save_results_path: str, 
+            log_progress: bool = True,
+            num_hpo_trials: int = None,
+        ):
+        self.gate_q_env_config = gate_q_env_config
+        self.q_env = QuantumEnvironment(self.gate_q_env_config)
```
[the remark this answers is not in the document]
Comment (lukasvoss) [reply hyperparameter_optimization.py]: Fixed. I now pass a QuantumEnvironment object to the HPO class. One can also pass a ContextAwareQuantumEnvironment object.

File: hyperparameter_optimization.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        assert isinstance(self.num_hpo_trials, int) and self.num_hpo_trials > 0, "num_hpo_trials must be an integer greater than 0"
+
+        
+    def _objective(self, trial):
+        # Fetch hyperparameters from the trial object
+        self.agent_config, self.hyperparams = create_agent_config(trial, self.hpo_config, self.network_config, self.ppo_params)
+
+        self.q_env = QuantumEnvironment(self.gate_q_env_config)
+        # Overwrite the batch_size of the environment with the one from the agent_config
+        self.q_env.batch_size = self.agent_config['BATCHSIZE']
```
[the remark this answers is not in the document]
Comment (lukasvoss) [reply hyperparameter_optimization.py:50]: The batchsize will be _dynamically_ calculated by introducing a given set of choosable minibatch_sizes and an integer-spectrum of multipliers for the HPO algo to choose from. This way, we can ensure that batchsize = k * mini_batchsize with $k \in \mathbb{N}^+$

File: template_configurations/qiskit/q_env_config.py (hunk 2 above)
Comment (lukasvoss) [reply template_configurations/qiskit/q_env_config.py:64]: You are right, I removed the factor 2.

File: quantumenvironment.py (hunk 1 above)
Comment (lukasvoss) [reply quantumenvironment.py:42]: Why is that? 
All three imports: ``DensityMatrix``, ``Statevector`` and ``Layout`` are being used in the quantumenvironment.py script.

Comment (arthurostrauss) [reply quantumenvironment.py:42]: I meant the qiskit.algorithms ComputeUncompute one, I must have missed the line.

Comment (lukasvoss) [reply quantumenvironment.py]: Removed it.

**Comment by lukasvoss** [comment]
I have implemented your suggestions.
These are the additional changes made 

- Added more documentation to the hpo wrapper
- The hpo now records and saves more trainings metrics
- The hpo has now an own dedicated yaml file
- All yaml configuration files are now in a separate directory

File: template_configurations/qiskit/q_env_config.py [hunk 5]
```diff
@@ -14,20 +15,21 @@
...
-from qiskit_ibm_runtime.fake_provider import FakeProvider
+from qiskit.providers.fake_provider import FakeProvider
 from qiskit.providers import BackendV1, BackendV2
 from qiskit.providers.fake_provider import FakeJakartaV2
 from qiskit_experiments.calibration_management import Calibrations
 from qconfig import QiskitConfig, QEnvConfig
 from quantumenvironment import QuantumEnvironment
 from context_aware_quantum_environment import ContextAwareQuantumEnvironment
-from dynamics_config import dynamics_backend
+from template_configurations.qiskit.dynamics_config import dynamics_backend
```
Comment (arthurostrauss) [inline template_configurations/qiskit/q_env_config.py:27]: You should not have to go through the full thing here either. Actually in this config we only care about the gate level anyway

File: template_configurations/qiskit/template_execution.ipynb
Comment (arthurostrauss) [inline template_configurations/qiskit/template_execution.ipynb:1]: Why do you want to delete it for now?

**Review by arthurostrauss — CHANGES_REQUESTED** [review]
Overall, few things.

You are changing some imports but I don't see why you have to do them.

Also, the yaml files are very tedious to review, and I would like to separate the different things. So my suggestion is: leave the current yaml files intact (with the current values on my branch), but just add an additional file for the hyper parameter optimisation on top of the others. 

Finally, don't delete any notebooks which serve for now as examples, I will refactor them when it's time.

File: template_configurations/qiskit/template_execution.ipynb
Comment (lukasvoss) [reply template_configurations/qiskit/template_execution.ipynb:1]: Reuploaded.

File: template_configurations/qiskit/q_env_config.py [hunk 6]
```diff
@@ -14,20 +15,21 @@
 from qiskit import pulse, QuantumCircuit, QuantumRegister, transpile
 from qiskit.circuit import ParameterVector, Gate
 from qiskit_dynamics import Solver, DynamicsBackend
-from custom_jax_sim import JaxSolver
+from custom_jax_sim.jax_solver import JaxSolver
```
[the remark this answers is not in the document]
Comment (lukasvoss) [reply template_configurations/qiskit/q_env_config.py]: Fixed.

File: template_configurations/qiskit/q_env_config.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 from typing import Optional, Dict
 import os
+import sys
```
[the remark this answers is not in the document]
Comment (lukasvoss) [reply template_configurations/qiskit/q_env_config.py]: Removed unused imports.

**Comment by lukasvoss** [comment]
Implemented the requested changes:

- Agent Config yaml file is now one block **without** sub-division in agent and network
- The helper functions in the helper_funtions.py script are updated accordingly
- The import statements are now leaner using the init files of the relevant directories
- Now the optimizer module from the hpo wrapper takes the input num_trials instead of when initialiazing the hpo wrapper object
- Added handling of overwriting the batchsize (so that it can be used as a hyperparameter) with the _unwrapped_ statement, otherwise the overwriting would not work properly
- Added handling in case all trials fail (due to a hyperparam configuration that leads to e.g., nan in the training tensors -> training fails -> expect block logs that this iteration fails, penalizes the hyperparam config and continues with the next trial)
- Now maximizing fidelity instead of reward

File: context_aware_quantum_environment.py [hunk 8]
```diff
@@ -32,7 +32,7 @@
 from qiskit_aer.backends.aerbackend import AerBackend
 from qiskit_aer.noise import NoiseModel
 from qiskit_aer.primitives import Estimator as AerEstimator, Sampler as AerSampler
-from qiskit_algorithms.state_fidelities import ComputeUncompute
+from qiskit.algorithms.state_fidelities import ComputeUncompute
```
Comment (arthurostrauss) [inline context_aware_quantum_environment.py:35]: Should come back to qiskit_algorithms and not qiskit.algorithms.

**Review by arthurostrauss — APPROVED** [review]

**Comment by arthurostrauss** [comment]
Seems that there are still some conflicts preventing me to merge. Can you rebase and figure them out?
